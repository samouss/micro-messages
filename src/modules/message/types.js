// @flow

export type MessageId = string;
export type MessageBody = string;

export type Visibility = 'public' | 'private';

export type Message = {
  +id: MessageId,
  +date: string,
  +body: MessageBody,
  +visibility: Visibility,
};

export type ByIdState = { [id: MessageId]: Message };
export type IdsState = Array<MessageId>;

export type MessageState = {
  +messages: {
    +byId: ByIdState,
    +ids: IdsState,
  },
};

export type MessageAction =
  | { type: 'FETCH_MESSAGES_REQUEST' }
  | { type: 'FETCH_MESSAGES_SUCCESS', messages: Array<Message> }
  | { type: 'FETCH_MESSAGES_FAILURE' };
