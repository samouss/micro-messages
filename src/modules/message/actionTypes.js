// @flow

export const FETCH_MESSAGES_REQUEST = 'FETCH_MESSAGES_REQUEST';
export const FETCH_MESSAGES_SUCCESS = 'FETCH_MESSAGES_SUCCESS';
export const FETCH_MESSAGES_FAILURE = 'FETCH_MESSAGES_FAILURE';
