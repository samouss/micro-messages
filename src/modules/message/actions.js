// @flow

import { fetchMessages as fetchMessagesFromApi } from 'core/api';
import type { Dispatch } from 'store/types';
import type { Message, MessageAction } from './types';

const fetchMessagesRequest = (): MessageAction => ({
  type: 'FETCH_MESSAGES_REQUEST',
});

const fetchMessagesSuccess = (messages: Array<Message>): MessageAction => ({
  type: 'FETCH_MESSAGES_SUCCESS',
  messages,
});

const fetchMessagesFailure = (): MessageAction => ({
  type: 'FETCH_MESSAGES_FAILURE',
});

export const fetchMessages = () => (dispatch: Dispatch) => {
  dispatch(fetchMessagesRequest());

  return fetchMessagesFromApi().then(res => {
    dispatch(fetchMessagesSuccess(res));
  }).catch(() => {
    dispatch(fetchMessagesFailure());
  });
};
