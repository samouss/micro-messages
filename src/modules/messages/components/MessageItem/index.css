.MessageItem {
  display: block;
}

.MessageItem:hover {
  background-color: #F0F0F0;
}

.MessageItem__Container {
  display: flex;
  flex-direction: row;
  height: 95px;
  padding: 5px 8px;
  border-bottom: 1px solid #E5E5E5;
}

.MessageItem__Preview {
  flex: 1;
  overflow: hidden;
}

.MessageItem__Infos {
  display: flex;
  flex-direction: column;
  margin-left: 10px;
  align-items: flex-end;
}

.MessageItem__Infos__Date {
  font-size: 0.9rem;
  white-space: nowrap;
}

.MessageItem__Infos__Visibility {
  margin-top: 6px;
}
