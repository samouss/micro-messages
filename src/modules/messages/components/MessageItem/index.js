// @flow

import React from 'react';
import { NavLink } from 'react-router-dom';
import TimeFromNow from 'components/TimeFromNow';
import Label from 'components/Label';
import Paragraphs from 'components/Paragraphs';
import './index.css';

import type { Message } from '../../types';

export type Props = {
  message: Message,
};

const MessageItem = ({ message }: Props) => (
  <li>
    <NavLink
      to={`/${message.id}`}
      styleName="MessageItem"
      activeStyle={{
        backgroundColor: '#0169D9',
        color: '#FFFFFF',
      }}
    >
      <div styleName="MessageItem__Container">
        <div styleName="MessageItem__Preview">
          <Paragraphs
            text={message.body}
          />
        </div>

        <div styleName="MessageItem__Infos">
          <span styleName="MessageItem__Infos__Date">
            <TimeFromNow
              date={message.date}
            />
          </span>

          <span styleName="MessageItem__Infos__Visibility">
            <Label
              type={message.visibility}
            />
          </span>
        </div>
      </div>
    </NavLink>
  </li>
);

export default MessageItem;
