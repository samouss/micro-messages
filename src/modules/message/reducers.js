// @flow

import { combineReducers } from 'redux';
import * as ACTION_TYPES from './actionTypes';
import type { Action } from 'store/types';
import type { MessageByIdState, MessageIdsState } from './types';

const messageByIdInitialState: MessageByIdState = {};
const messageById = (state = messageByIdInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return {
        ...state,
        ...action.messages.reduce((acc, message) => ({
          ...acc,
          [message.id]: message,
        }), {}),
      };
    }

    default: {
      return state;
    }
  }
};

const messageIdsInitialState: MessageIdsState = [];
const messageIds = (state = messageIdsInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return state.concat(action.messages.map(message => message.id));
    }

    default: {
      return state;
    }
  }
};

export default combineReducers({
  messageById,
  messageIds,
});
