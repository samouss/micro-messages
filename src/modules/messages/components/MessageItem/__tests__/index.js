// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { createMockMessage } from 'test/message';
import MessageItem from '../index';

import type { Props } from '../index';

describe('<MessageItem />', () => {
  it('expect to render the component correctly', () => {
    const props: Props = {
      message: createMockMessage('ID1234'),
    };

    const component = shallow(
      <MessageItem
        {...props}
      />,
    );

    expect(component).toMatchSnapshot();
  });
});
