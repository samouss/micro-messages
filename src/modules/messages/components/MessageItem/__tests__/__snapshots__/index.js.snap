// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageItem /> expect to render the component correctly 1`] = `
<li>
  <NavLink
    activeClassName="active"
    activeStyle={
      Object {
        "backgroundColor": "#0169D9",
        "color": "#FFFFFF",
      }
    }
    styleName="MessageItem"
    to="/ID1234"
  >
    <div
      styleName="MessageItem__Container"
    >
      <div
        styleName="MessageItem__Preview"
      >
        <Paragraphs
          text="BODY"
        />
      </div>
      <div
        styleName="MessageItem__Infos"
      >
        <span
          styleName="MessageItem__Infos__Date"
        >
          <TimeFromDate
            date="DATE"
          />
        </span>
        <span
          styleName="MessageItem__Infos__Visibility"
        >
          <Label
            type="private"
          />
        </span>
      </div>
    </div>
  </NavLink>
</li>
`;
